Add Slab, Time, and time steppers

wthrowe opened the pull request:
Notable changes:
* This has all been moved to a new Time directory which will later also contain the step-changing stuff.
* The interface for the main TimeStepper method has changed to something closer to what I want for local time stepping.
* Lots of functions were moved out of classes or into cpp files.
* The time stepper tests were completely redone, and I cleaned up the Time tests and added a few more cases.

wthrowe committed:
Add Time and Slab

wthrowe committed:
Add TimeSteppers

osheamonn commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,125 @@
+class er;
+}
+
+/// A chunk of time.  Every element must reach slab boundaries
```
Do we want this to be in some Doxygen group?

osheamonn commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,125 @@
+inline bool operator>=(const Slab& a, const Slab& b) { return not(a < b); }
+//@}
+
+// Output
```
Redundant comment

mar-celine commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,125 @@
+class TimeDelta;
+
+namespace PUP {
+class er;
```
what is class er?

osheamonn replied:
The thing which pup's is 'er' in the namespace PUP, so when you use it it's PUP::er...

nilsdeppe replied:
It's the serialization class used by Charm++. I don't think we need elaboration on it in the code since it's inside the `PUP` namespace, though I'd be find with a comment along the lines of `// Charm++` somewhere here too

mar-celine commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,125 @@
+  /// (up to roundoff) duration.
+  Slab retreat() const { return Slab(start_ - (end_ - start_), start_); }
+
+  /// Create a slab adjecent to this one in the direction indicated by
```
`adjacent` (in time?)

mar-celine commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,254 @@
+  }
+
+  Slab slab() const { return slab_; }
+  rational_t fraction() const { return fraction_; }
```
'length_of_fraction'? unless this returns a number between 0 and 1

mar-celine replied:
basing my comment off the comment in line 91

wthrowe replied:
The comment on line 91 isn't about this function.  I don't understand.

mar-celine replied:
sorry! misread

mar-celine commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,254 @@
+std::ostream& operator<<(std::ostream& os, const TimeDelta& dt);
+
+// Time <cmp> Time
+bool operator==(const Time& a, const Time& b);
```
are `noexcept`s needed anywhere here

wthrowe replied:
What is our policy on where to put `noexcept`?  I know it's important for things like move constructors.  As far as I know none of our code (except for a few places dealing with external libraries) is expected to throw or catch exceptions, so should we just mark every function with it?  Something in between?

nilsdeppe replied:
I think we should follow the gaming industry almost. They completely compile without exceptions but that's not an option for us. From everything I've been able to find we should `noexcept` everything we know won't throw, which is almost everything for us since we are stuck with terminate in almost all cases where we handle errors

mar-celine commented on src/Time/TimeSteppers/AdamsBashforthN.cpp:
```diff
@@ -0,0 +1,142 @@
+    const std::vector<double>& steps) {
+  const size_t order = steps.size();
+
+  // polynomials[coefficient][term exponent]
```
Can polynomials / integrals of polynomials be put in some Numerical like place

wthrowe replied:
At the very least I should add some comments here explaining what this is doing.  Once I've looked up what that is we can see whether it makes sense where it is.

nilsdeppe replied:
Unless this happens to be something general like Jacobi polynomials I think a reference + an equation is sufficient, just so that if someone in the future tries to make improvements on anything by using different formulas they can find the original.

mar-celine commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -0,0 +1,188 @@
+  // using OptionsList = tmpl::list<TargetOrder, SelfStart>;
+  // static std::string class_id() { return "AdamsBashforthN"; }
+  // static constexpr OptionString_t help = {
+  //     "An Adams-Bashforth Nth order time-stepper. The target order is the "
```
add blank line

wthrowe replied:
Where?  This comment is in the middle of a initializer.

nilsdeppe replied:
Let's leave this as is. We can deal with formatting preferences when it's uncommented for Factory.

mar-celine commented on src/Time/TimeSteppers/RungeKutta3.hpp:
```diff
@@ -0,0 +1,93 @@
+/// Galerkin Methods. section 5.7
+class RungeKutta3 : public TimeStepper::Inherit {
+ public:
+  //using OptionsList = tmpl::list<>;
```
will this be uncommented later

wthrowe replied:
Factory stuff

mar-celine commented on src/Time/TimeSteppers/RungeKutta3.hpp:
```diff
@@ -0,0 +1,93 @@
+
+  switch (substep) {
+    case 0: {
+      // from (5.32) of Hesthaven
```
latex/doxygen comment

wthrowe replied:
Do Doxygen comments work on random pieces of code?

nilsdeppe replied:
In this case I don't see what it would bind to, so no. As I see it, we have three options: 
1) leave as is
2) move all this type of documentation onto the function giving a nice LaTeX documentation explaining how the method works.
3) Same as two but put the Doxygen comment on the class

I don't have a strong preference here yet, but will think about it more. I'm mostly inclined to vote for requiring 3) but not denying 1) since there is merit for documentation in code. Having said that, Google's guide is that algorithm documenting documentation should be before the tricky bits of code, not dispersed throughout. Since I think here the documentation will be more likely to be read if on the class, I think that would be the most logical place. Just my current thoughts on the manner, let's have a discussion here about it.

wthrowe replied:
I think these would be hard to move out-of-line and have them be as useful, since they are mostly explaining what the states of the variables are at different points in the function.

mar-celine replied:
My motivation for making this comment was the lines of commented equations that follow - if there is a way to see a LaTeX version of those equations somewhere, with the class documentation, it should probably be available, is what I'm saying.

nilsdeppe replied:
Okay, so I agree with Will that they serve a good purpose inline. I also agree with Gabe that it would be useful (and maybe this is already there, in that case ignore) to have a description of the algorithm in documentation on the class. This is what I ended up doing for the p-adaptivity classes. Specifically, I think one or more references, whatever is useful, would be good to have. However, since the reference can be a book and so somebody might not have access to the book, having the relevant equations would also be valuable. I do not think we need to derive any equations. If there's a simple intuitive reason, that could be good to have, but in the case of integrators where the reason is "lots of math" just the equations for the steps is plenty. Thoughts from others?

wthrowe replied:
The current class documentation (author unknown to me) is
> A "strong stability-preserving" 3rd-order Runge-Kutta time-stepper.
> Major reference:  J. Hesthaven & T. Warburton, Nodal Discontinuous
> Galerkin Methods. section 5.7

This isn't a method I've studied at all, and I don't know the reasoning behind what it is doing.

mar-celine commented on tests/Unit/Time/Test_Slab.cpp:
```diff
@@ -0,0 +1,214 @@
+}
+
+// [[OutputRegex, Cannot compare overlapping slabs]]
+TEST_CASE("Unit.Time.Slab.lt.1", "[Unit][Time]") {
```
`lt.1` might be hard to read - maybe `lesser` and `greater`?

mar-celine commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,485 @@
+        h(Time(Slab(0, 2), rational_t(1, 3))));
+}
+
+TEST_CASE("Unit.Time.Time_slab_comparison", "[Unit][Time]") {
```
casing

mar-celine commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,485 @@
+        Time(slab.retreat(), rational_t(2, 3)));
+}
+
+TEST_CASE("Unit.Time.TimeDelta.serialization", "[Unit][Time][Serialization]") {
```
casing

mar-celine replied:
Actually I'm not sure what the rules are for test casing

mar-celine commented on tests/Unit/Time/TimeSteppers/TimeStepperTestUtils.hpp:
```diff
@@ -0,0 +1,182 @@
+  std::deque<std::tuple<Time, double, double>> history;
+
+  if (not stepper.is_self_starting()) {
+    initialize_history(time, history, stepper, analytic, rhs, step_size);
```
`make_history`? just kidding

wthrowe commented:
The test failures are due to that old SpEC topic of discussion: the [fused multiply-add](https://en.wikipedia.org/wiki/Multiply%E2%80%93accumulate_operation#Fused_multiply.E2.80.93add).

In this case the level of the floating-point noise floor was arbitrary, so I can just bump it up a bit to cover the very slightly higher error in the FMA version and everything will be fine, but we should consider whether this sort of thing worries us in general.

wthrowe commented:
Addressed most of the simple comments, hopefully fixed Travis failures.

Also changed the Time and TimeDelta output format, which I had meant to do before submitting.

Codecov seems confused.  It is complaining about files the pull request doesn't touch.

kidder commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,264 @@
+
+  /// A time a given fraction of the way through the given slab.
+  Time(Slab slab, rational_t fraction) noexcept
+      : slab_(slab), fraction_(fraction) {
```
could std::move(slab) to future-proof in case a Slab becomes less trivial

kidder commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,264 @@
+
+  /// An interval covering a given fraction of the slab.
+  TimeDelta(Slab slab, rational_t fraction) noexcept
+      : slab_(slab), fraction_(fraction) {}
```
could std::move(slab) to future-proof in case a Slab becomes less trivial

kidder commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,264 @@
+
+  /// Move the time to a different slab.  The time must be at an end
+  /// of the current slab and the new slab must share that endpoint.
+  Time with_slab(const Slab& slab) const noexcept;
```
how about create_with_slab which makes it clearer that a new Time is created instead of replacing the slab of the current object

wthrowe replied:
As with the comments on Slab, I do not like the new suggested name here.

kidder commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,264 @@
+}
+class TimeDelta;
+
+/// The time in a simulation.  Times can be safely compared for exact
```
Let's create a doxygen group named 'TimeGroup' (so there is no conflict with Time) and in GroupDefs.hpp add an overall comment explaining how Time, Slab, and TimeDelta fit together

kidder commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,264 @@
+  /// Move the interval to a different slab.  The resulting interval
+  /// will in general not be the same length, but will take up the
+  /// same fraction of its slab.
+  TimeDelta with_slab(const Slab& slab) const noexcept {
```
create_with_slab (see above)

kidder commented on src/Time/Time.hpp:
```diff
@@ -0,0 +1,264 @@
+
+  Slab slab() const noexcept { return slab_; }
+  rational_t fraction() const noexcept { return fraction_; }
+  /// Approximate numerical length of the interval.
```
blank line before comment, please

kidder commented on src/Time/Time.cpp:
```diff
@@ -0,0 +1,146 @@
+  }
+}
+
+Time Time::with_slab(const Slab& slab) const noexcept {
```
I would rename slab to new_slab to make things clearer

kidder commented on src/Time/Time.cpp:
```diff
@@ -0,0 +1,146 @@
+  if (t.is_at_slab_boundary()) {
+    return hash<double>()(t.value());
+  } else {
+    return hash<Slab>()(t.slab()) ^ hash<ssize_t>()(t.fraction().numerator()) ^
```
xor instead of ^ ?

kidder commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,137 @@
+
+  /// Construct a slab running between two times (exactly).
+  Slab(double start, double end) noexcept : start_(start), end_(end) {
+    ASSERT(start_ < end_, "Backwards Slab");
```
can the error message make clearer whether this is a temporary restriction or a permanent one (i.e. will backwards slabs ever be supported?)

wthrowe replied:
I don't intend to support them, but it seems weird to promise that something will never be supported unless I have a good reason that it can't work.  Suggestions for what sort of message I should have?

kidder commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,137 @@
+
+  /// Construct a slab with a given start time and duration.  The
+  /// actual duration may differ by roundoff from the supplied value.
+  static Slab new_with_duration(double start, double duration) noexcept {
```
with_duration_from_start

wthrowe replied:
This replies to a few comments:

The overall suggestions are `new_with_*` -> `with_*` and `with_*` -> `create_with_*`.  (The other portions of the suggestions I like and will change.)  This is reversing my original naming scheme and seems very confusing to me (new and create are approximately synonyms).  In particular, I expect methods starting with "create" to be factory-like, and either be static methods or methods on some sort of factory object.

`new_with_` -> `with_` seems okay, except that it depends on the change I don't like.

kidder replied:
Okay, so for the static function I felt the new was not needed as it obviously had to make a new object.  I have a slight concern about the member variable with_* being misused by someone thinking it modifies the object instead of creating a new one, hence my suggestion of create_with_*, but I understand your point about this naming being the opposite of our factory-like convention.  Since they both create a new object and have different interfaces, I would be okay with them sharing the same name.  Anyone else have an opinion?

kidder commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,137 @@
+
+  /// Construct a slab with a given end time and duration.  The
+  /// actual duration may differ by roundoff from the supplied value.
+  static Slab new_with_duration_reverse(double end, double duration) noexcept {
```
with_duration_to_end

kidder commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,137 @@
+
+  /// Create a new slab with the same start time as this one with the
+  /// given duration (up to roundoff).
+  Slab with_duration(double duration) const noexcept {
```
create_with_duration_from_start

kidder commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,137 @@
+
+  /// Create a new slab with the same end time as this one with the
+  /// given duration (up to roundoff).
+  Slab with_duration_reverse(double duration) const noexcept {
```
create_with_duration_to_end

kidder commented on src/Time/Slab.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace std {
+size_t hash<Slab>::operator()(const Slab& s) const noexcept {
+  return hash<double>()(s.start().value()) ^ hash<double>()(s.end().value());
```
xor instead of ^ ?

wthrowe replied:
I don't really like the keywords for the bitwise operations, mostly because they're named really terribly.  `and` is logical, but `xor` and `and_eq` are bitwise and similar to `bitand`.

This should probably be using the boost hash combining functions anyway, though, so I'll do that.

kidder replied:
I think I agree with you, I'd rather use the symbols for bitwise operators (Nils raised the question in code review by left it as soliciting your opinion, not a requested change)

kidder commented on src/Time/Slab.hpp:
```diff
@@ -0,0 +1,137 @@
+/// slabs should not be compared (and will trigger an assertion).
+//@{
+inline bool operator<(const Slab& a, const Slab& b) noexcept {
+  ASSERT(a == b or a.end_ <= b.start_ or a.start_ >= b.end_,
```
why is checking < with yourself an okay operation?  
If it is okay, then the comment should say so

wthrowe replied:
I'll update the comment, and I'll make sure to include the reasoning behind this in the overall Slab/Time explanation for the doxygen group.

kidder commented on tests/Unit/Time/Test_Slab.cpp:
```diff
@@ -0,0 +1,277 @@
+  const Slab slab(tstart_d, tend_d);
+
+  // Arbitrary
+  const double tend2_d = tend_d + 1.234;
```
In some of the tests below, tend2_d is used as a duration instead of the end of a slab; I would suggest using separate names (you keep the same value )

kidder commented on tests/Unit/Time/Test_Slab.cpp:
```diff
@@ -0,0 +1,277 @@
+  }
+
+  {
+    const Slab slab2 = Slab::new_with_duration(tstart_d, tend_d);
```
Just to make things clearer, use another variable instead of tend_d for a duration

kidder commented on tests/Unit/Time/Test_Slab.cpp:
```diff
@@ -0,0 +1,277 @@
+  ASSERTION_TEST();
+#ifdef SPECTRE_DEBUG
+  Slab(1., 0.);
+  ERROR("Bad end");
```
I know this was copied from Nils, but replace "Bad end" with "The previous line should have triggered an ASSERT"

wthrowe replied:
The current form is mandated by our dev guide.  Please change it there first if it is going to be changed.

kidder replied:
okay, I will do that

kidder commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,606 @@
+
+TEST_CASE("Unit.Time.Time.serialization", "[Unit][Time][Serialization]") {
+  const Slab slab(1.7, 2.4);
+  const Time time = slab.start() + slab.duration() * 3 / 5;
```
I prefer 3./5. (to show good examples for new programmers, or someone reorders the operands)

wthrowe replied:
More exact math.

kidder commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,606 @@
+
+  CHECK(TimeDelta(slab, 0).value() == 0);
+  CHECK(TimeDelta(slab, 1).value() == approx(length_d));
+  CHECK(TimeDelta(slab, rational_t(1, 5)).value() == approx(length_d / 5));
```
5.

wthrowe replied:
more exact math

kidder commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,606 @@
+  CHECK(TimeDelta(slab, 1).value() == approx(length_d));
+  CHECK(TimeDelta(slab, rational_t(1, 5)).value() == approx(length_d / 5));
+  // TimeDeltas can be "out of range".
+  CHECK(TimeDelta(slab, -rational_t(1, 5)).value() == approx(-length_d / 5));
```
5.

wthrowe replied:
again

kidder commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,606 @@
+  const Slab slab(tstart_d, tend_d);
+  CHECK(TimeDelta(slab, 0).slab() == slab);
+  CHECK(TimeDelta(slab, rational_t(3, 5)).fraction() == rational_t(3, 5));
+  CHECK(TimeDelta(slab, rational_t(3, 5)).fraction() == rational_t(3, 5));
```
identical to previous line

kidder commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,606 @@
+
+TEST_CASE("Unit.Time.TimeDelta.serialization", "[Unit][Time][Serialization]") {
+  const Slab slab(1.7, 2.4);
+  const TimeDelta dt = slab.duration() * 3 / 5;
```
3./5.

wthrowe replied:
again

kidder commented on tests/Unit/Time/Test_Time.cpp:
```diff
@@ -0,0 +1,606 @@
+  ASSERTION_TEST();
+#ifdef SPECTRE_DEBUG
+  Time(Slab(0., 1.), -1);
+  ERROR("Bad end");
```
whatever I said above about Bad end

kidder commented on src/Time/TimeSteppers/TimeStepper.hpp:
```diff
@@ -0,0 +1,71 @@
+#include "Utilities/FakeVirtual.hpp"
+
+/// Holds classes that take time steps.
+namespace TimeSteppers {}
```
remove this and the following blank line

kidder commented on src/Time/TimeSteppers/TimeStepper.hpp:
```diff
@@ -0,0 +1,71 @@
+DEFINE_FAKE_VIRTUAL(update_u)
+}  // TimeStepper_detail
+
+/// Abstract base class for TimeSteppers.
```
create a doxygen group TimeSteppersGroup and add that this is in the group

kidder commented on src/Time/TimeSteppers/TimeStepper.hpp:
```diff
@@ -0,0 +1,71 @@
+  /// \endcond
+
+  /// Add the change for the current substep to u and return the substep size.
+  /// More recent values should be later in the history list.
```
replace with /// The history is chronological order.

wthrowe replied:
It's not in chronological order in general.  For AB running backwards it's reverse chronological, for RK3 it jumps around.

kidder replied:
Then I don't understand what 'more recent values' means; should we use a queue instead of a deque?

wthrowe replied:
More recent in wall clock time, or something along those lines.  The statement is trying to say it's being used in a queue-like manner and specifying which end to push things onto.  Unfortunately, `std::queue` is missing most of the container operations and so is unusable here, even though logically it would be clearer.

kidder replied:
Okay, I hadn't realized that std:queue was so pathetic (I thought it just removed the deque functions to enforce FIFO).  Maybe we could just say it should be used as a queue (i.e. push back pop front)

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -0,0 +1,188 @@
+/// An Nth Adams-Bashforth time stepper.
+class AdamsBashforthN : public TimeStepper::Inherit {
+ public:
+  static constexpr const size_t MAXIMUM_ORDER = 8;
```
maximum_order

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -0,0 +1,188 @@
+std::vector<double> AdamsBashforthN::get_coefficients(
+    const Iterator& times_begin, const Iterator& times_end,
+    const TimeDelta& step) noexcept {
+  std::vector<double> steps;
```
reserve space for vector

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -0,0 +1,188 @@
+         "Length of history (" << history.size() << ") "
+         << "should not exceed target order (" << target_order_ << ")");
+
+  const Time& (*get_time)(const decltype(history[0])&) = std::get<0>;
```
add a comment for the C++ neophytes like Francois 😛

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.cpp:
```diff
@@ -0,0 +1,142 @@
+    }
+  }
+
+  std::vector<double> denominators;
```
reserve known size

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.cpp:
```diff
@@ -0,0 +1,142 @@
+    denominators.push_back(denom);
+  }
+
+  std::vector<double> result;
```
reserve known size

kidder commented on src/Time/TimeSteppers/RungeKutta3.hpp:
```diff
@@ -0,0 +1,93 @@
+      // On entry V = v^(1), U0 = u^n, rhs0 = RHS(v^(1),t^n + dt),
+      // time = t^n + dt
+      u += 0.25 * (3.0 * (U0 - vars) + step_size.value() * dt_vars);
+      return -step_size / 2;
```
return -0.5*step_size;

wthrowe replied:
This is an exact rational type.

kidder commented on src/Time/TimeSteppers/RungeKutta3.hpp:
```diff
@@ -0,0 +1,93 @@
+      // On entry V = v^(2), U0 = u^n, rhs0 = RHS(v^(2),t^n + (1/2)*dt),
+      // time = t^n + (1/2)*dt
+      u += (1.0 / 3.0) * (U0 - vars + 2.0 * step_size.value() * dt_vars);
+      return step_size / 2;
```
return 0.5*step_size;

wthrowe replied:
exact math

kidder commented on tests/Unit/Time/TimeSteppers/TimeStepperTestUtils.hpp:
```diff
@@ -0,0 +1,183 @@
+
+#include <catch.hpp>
+#include <cmath>
+#include <tuple>
```
`#include <deque>`

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -0,0 +1,188 @@
+
+#include <array>
+#include <boost/iterator/transform_iterator.hpp>
+#include <iterator>
```
`#include <deque>`

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -0,0 +1,188 @@
+#include <iterator>
+#include <memory>
+#include <string>
+#include <vector>
```
`#include <tuple>`

kidder commented on src/Time/TimeSteppers/RungeKutta3.hpp:
```diff
@@ -0,0 +1,93 @@
+
+#include <deque>
+#include <string>
+
```
`#include <tuple>`

kidder commented on src/Time/TimeSteppers/TimeStepper.hpp:
```diff
@@ -0,0 +1,71 @@
+  virtual ~TimeStepper() noexcept = default;
+  /// \endcond
+
+  /// Add the change for the current substep to u and return the substep size.
```
Clarify that the substep size is the increment (or decrement) in the time at which the next substep evaluates its RHS

kidder reviewed (changes requested):
These are from a 622 code review also with Nils, Gabe, Francois, Mark and Eamonn.  We still need look at the tests for the time steppers

kidder reviewed (changes requested):
squash and push

wthrowe commented:
There are a couple of valid warnings in the 2000+ lines of clang-tidy spew.
